src/parse/message.rs

```diff
@@ -1,219 +1,172 @@
-fn read_headers(reader: &mut impl BufRead) -> Result<HeaderMap, ParsingError> {
-    let mut reader = reader.error_take(MAX_HEADERS_LENGTH);
-
-    let mut headers = header_map![];
-
-    loop {
-        let line = read_line(&mut reader)?;
+/// Parses the first line and returns the next state if possible.
+fn first_line_state<T, R: Parse<T>>(reader: &mut impl BufRead, parser: R) -> ParseResult<State<R, T>, State<R, T>> {
+    Ok(match parser.parse(reader)? {
+        Done(first_line) => Done(Headers(first_line, HeadersParser::new())),
+        IoErr(parser, err) => IoErr(FirstLine(parser), err)
+    })
+}
 
-        if line.is_empty() {
-            return Ok(headers);
+/// Parses the headers and returns the next state if possible.
+fn headers_state<T, R>(reader: &mut impl BufRead, first_line: T, parser: HeadersParser, read_body_if_no_content_length: bool) -> ParseResult<State<R, T>, State<R, T>> {
+    Ok(match parser.parse(reader)? {
+        Done(headers) => {
+            let body_parser = BodyParser::new(&headers, read_body_if_no_content_length)?;
+            Done(Body(first_line, headers, body_parser))
         }
-
-        let (header, value) = parse_header(line)?;
-        headers.add_header(header, value);
-    }
+        IoErr(parser, err) => IoErr(Headers(first_line, parser), err)
+    })
 }
 
-/// Parses the given line as a header. Splits the line at the first ": " pattern.
-fn parse_header(raw: String) -> Result<(Header, String), ParsingError> {
-    let mut split = raw.splitn(2, ": ");
-
-    let header_raw = split.next().ok_or(ParsingError::BadSyntax)?;
-    let value = split.next().ok_or(ParsingError::BadSyntax)?;
-
-    Ok((Header::from(header_raw), String::from(value)))
+/// Parses the body and returns the next state if possible.
+fn body_state<T, R>(reader: &mut impl BufRead, first_line: T, headers: HeaderMap, parser: BodyParser) -> ParseResult<State<R, T>, State<R, T>> {
+    Ok(match parser.parse(reader)? {
+        Done(body) => Done(Finished(first_line, headers, body)),
+        IoErr(parser, err) => IoErr(Body(first_line, headers, parser), err)
+    })
 }
 
+
 #[cfg(test)]
 mod tests {
-    use std::io::{BufReader, Error, ErrorKind};
+    use std::io::{Error, ErrorKind};
 
     use crate::common::header::{CONTENT_LENGTH, Header, HeaderMap, HeaderMapOps, TRANSFER_ENCODING};
-    use crate::parse::common::read_message;
-    use crate::parse::error::ParsingError;
-    use crate::parse::error::ParsingError::{BadSyntax, EOF, InvalidChunkSize, InvalidHeaderValue, Reading};
-    use crate::util::mock::{EndlessMockReader, MockReader};
+    use crate::header_map;
+    use crate::parse::crlf_line::CrlfLineParser;
+    use crate::parse::error::ParsingError::{BadSyntax, InvalidChunkSize, InvalidHeaderValue};
+    use crate::parse::message::MessageParser;
+    use crate::parse::test_util;
+    use crate::parse::test_util::TestParseResult;
+    use crate::parse::test_util::TestParseResult::{ParseErr, Value};
```

Comment:
I noticed a lot of your tests import so much stuff. You can use super::* to avoid having to import everything (although I'm guessing you just import stuff automatically in Intellij)

Reply:
Ohhh yeah that's good to know. Yeah I've just been auto-importing with intellij haha.

Reply:
Yeah I figured lol